lib/ui/viewer/search/search_widget.dart

```diff
@@ -139,34 +143,63 @@ class _SearchWidgetState extends State<SearchWidget> {
     );
   }
 
-  Future<List<SearchResult>> getSearchResultsForQuery(String query) async {
-    final List<SearchResult> allResults = [];
+  @override
+  void dispose() {
+    if (_debounce != null) {
+      _debounce.cancel();
+    }
+    super.dispose();
+  }
 
-    final queryAsIntForYear = int.tryParse(query);
-    if (_isYearValid(queryAsIntForYear)) {
-      final yearResult =
-          await _searchService.getYearSearchResults(queryAsIntForYear);
-      allResults.add(yearResult); //only one year will be returned
+  Future getSearchResultsForQuery(String query) async {
+    final List<SearchResult> allResults = [];
+    if (query.isEmpty) {
+      if (_debounce != null && _debounce.isActive) {
+        _debounce.cancel();
+      }
+      return (allResults);
     }
 
-    final holidayResults = await _searchService.getHolidaySearchResults(query);
-    allResults.addAll(holidayResults);
+    final completer = Completer();
+
+    _debounceQuery(() async {
+      final queryAsIntForYear = int.tryParse(query);
+      if (_isYearValid(queryAsIntForYear)) {
+        final yearResult =
+            await _searchService.getYearSearchResults(queryAsIntForYear);
+        allResults.add(yearResult); //only one year will be returned
+      }
+
+      final holidayResults =
+          await _searchService.getHolidaySearchResults(query);
+      allResults.addAll(holidayResults);
 
-    final collectionResults =
-        await _searchService.getCollectionSearchResults(query);
-    allResults.addAll(collectionResults);
+      final collectionResults =
+          await _searchService.getCollectionSearchResults(query);
+      allResults.addAll(collectionResults);
 
-    final locationResults =
-        await _searchService.getLocationSearchResults(query);
-    allResults.addAll(locationResults);
+      final locationResults =
+          await _searchService.getLocationSearchResults(query);
+      allResults.addAll(locationResults);
 
-    final monthResults = await _searchService.getMonthSearchResults(query);
-    allResults.addAll(monthResults);
+      final monthResults = await _searchService.getMonthSearchResults(query);
+      allResults.addAll(monthResults);
 
-    return allResults;
+      completer.complete(allResults);
+    });
+
+    return completer.future;
   }
 
   bool _isYearValid(int year) {
     return year != null && year >= 1970 && year <= currentYear;
   }
+
+  _debounceQuery(fn) {
```

Comment:
You could extract this to a separate class, called `Debouncer`, that takes the `Duration` as a parameter? You could then keep calling `_debouncer.run(fn)`, and eventually `_debouncer.cancel()` on dispose.